### PR TITLE
Refactor sql analysis

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Refactor sql analysis

### Added

- General filter module for Analysis
- Tests

### Changed

- Refactor sql analysis and migrate it ReadHandler

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b refactor_sql_analysis -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: trailblazer/store/crud/read.py [hunk 1]
```diff
@@ -30,6 +31,15 @@ def get_nr_jobs_with_status_per_category(
...
         return [{"name": category.name, "count": category.count} for category in categories]
 
+    def get_analysis_with_id(self, analysis_id: int) -> Optional[Analysis]:
+        """Get a single analysis by id."""
+        return apply_analysis_filter(
+            filter_functions=[AnalysisFilter.FILTER_BY_ID],
+            analyses=self.get_query(table=Analysis),
+            analysis_id=analysis_id,
+        ).first()
+        return self.Analysis.query.get(analysis_id)
```
Comment (diitaz93) [inline trailblazer/store/crud/read.py:41]: I this second return supposed to be here?

**Review by diitaz93 — APPROVED** [review]
Neat ⭐ Just one concern about the second return

**Comment by henrikstranneheim** [comment]
No, was a leftover from the refactoring. Will remove

File: tests/store/filters/test_analysis_filters.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    """Test return analysis by id when existing."""
+    # GIVEN a store containing analyses
+    existing_analysis: Analysis = analysis_store.get_query(table=Analysis).first()
+
+    # WHEN retrieving an analysis by id
+    analysis: Query = filter_analyses_by_id(
+        analyses=analysis_store.get_query(table=Analysis), analysis_id=existing_analysis.id
+    )
+
+    # ASSERT that the analysis is a query
```
Comment (Vince-janv) [inline tests/store/filters/test_analysis_filters.py:18]: I'm not sure what the `ASSERT` keyword adds compared to GWT. Doesn't `THEN` also work here?

**Review by Vince-janv — APPROVED** [review]
Nice improvement 🔥

File: trailblazer/store/crud/read.py (hunk 1 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/read.py]: Fixed

File: tests/store/filters/test_analysis_filters.py (hunk 2 above)
Comment (henrikstranneheim) [reply tests/store/filters/test_analysis_filters.py]: Fixed

**Comment by henrikstranneheim** [comment]
### CLI ✅ 
```
θ70° [henrik.stranneheim@hasta:~] [S_base] 1m42s $ trailblazer scan
Status in SLURM: justhusky - 271984
          id               step     status time_limit  time_elapsed              started
0    4975989          fastqc_ar  COMPLETED   10:00:00            42  2023-07-30T14:35:59
1    4975990          fastqc_ar  COMPLETED   10:00:00            42  2023-07-30T14:35:59
2    4975991          fastqc_ar  COMPLETED   10:00:00            42  2023-07-30T14:35:59
3    4975995     samtools_merge    PENDING   10:00:00             0                  NaN
4    4975996     samtools_merge    PENDING   10:00:00             0                  NaN
..       ...                ...        ...        ...           ...                  ...
101  4976093  analysisrunstatus    PENDING    1:00:00             0                  NaN
102  4976094              sacct    PENDING    1:00:00             0                  NaN
103  4975994           bwa_mem2    RUNNING   10:00:00            55  2023-07-30T14:43:42
104  4975993           bwa_mem2    RUNNING   10:00:00            62  2023-07-30T14:36:55
105  4975992           bwa_mem2    RUNNING   10:00:00            62  2023-07-30T14:36:54

[106 rows x 6 columns]
Updated status justhusky - 271984: running
slurm_load_jobs error: Invalid job id specified
Error logging case - HYG7YDSXX : CalledProcessError
Error logging case - gentleamoeba : FileNotFoundError
Error logging case - masterimpala : FileNotFoundError
Error logging case - starakita : FileNotFoundError
Error logging case - goodrodent : FileNotFoundError
Error logging case - strongbug : FileNotFoundError
Error logging case - masterwombat : FileNotFoundError
Error logging case - warmcalf : FileNotFoundError
Error logging case - jointstag : FileNotFoundError
Error logging case - finerobin : FileNotFoundError
Error logging case - wantedhen : FileNotFoundError
Error logging case - cleanleech : PermissionError
Error logging case - socialearwig : FileNotFoundError
Error logging case - topsdodo : FileNotFoundError
Error logging case - usablefeline : PermissionError
Error logging case - 233471 : FileNotFoundError
Error logging case - crispshiner : FileNotFoundError
Error logging case - supertiger : PermissionError
Error logging case - fullfrog : FileNotFoundError
Error logging case - toplouse : FileNotFoundError
Error logging case - sunnycondor : FileNotFoundError
Error logging case - keyeft : FileNotFoundError
Error logging case - robustalien : PermissionError
Error logging case - busyowl : PermissionError
Error logging case - boldhound : PermissionError
Error logging case - vastboar : FileNotFoundError
Error logging case - topsworm : PermissionError
Error logging case - setcolt : PermissionError
Error logging case - proudfish : PermissionError
Error logging case - funnyaphid : FileNotFoundError
Error logging case - onestag : PermissionError
Error logging case - newgopher : PermissionError
Error logging case - helpedlab : EmptySqueueError
Error logging case - goodbengal : EmptySqueueError
Error logging case - grandaphid : PermissionError
All analyses updated!
```

### API ✅ 
Screenshot 2023-07-30 at 15 46 33

**Comment by henrikstranneheim** [comment]
## Cancel analysis ✅ 
```
θ69° [henrik.stranneheim@hasta:~] [S_base] 4s $ trailblazer cancel 271984
Cancelling job 4975995 - samtools_merge
Cancelling job 4975996 - samtools_merge
Cancelling job 4975997 - samtools_merge
Cancelling job 4975998 - markduplicates
Cancelling job 4975999 - markduplicates
Cancelling job 4976000 - markduplicates
Cancelling job 4976026 - cnvnator_ar
Cancelling job 4976027 - cnvnator_ar
Cancelling job 4976028 - cnvnator_ar
Cancelling job 4976030 - manta
Cancelling job 4976034 - sv_varianteffectpredictor
Cancelling job 4976035 - sv_vcfparser
Cancelling job 4976036 - sv_rankvariant
Cancelling job 4976049 - deepvariant
Cancelling job 4976050 - deepvariant
Cancelling job 4976051 - deepvariant
Cancelling job 4976052 - glnexus_merge
Cancelling job 4976068 - prepareforvariantannotationblock
Cancelling job 4976069 - rhocall_ar
Cancelling job 4976070 - bcftools_norm
Cancelling job 4976071 - variant_annotation
Cancelling job 4976081 - frequency_filter
Cancelling job 4976082 - cadd_ar
Cancelling job 4976083 - varianteffectpredictor
Cancelling job 4976084 - vcfparser_ar
Cancelling job 4976085 - rankvariant
Cancelling job 4976001 - chanjo_sexcheck
Cancelling job 4976002 - chanjo_sexcheck
Cancelling job 4976003 - chanjo_sexcheck
Cancelling job 4976004 - samtools_subsample_mt
Cancelling job 4976005 - samtools_subsample_mt
Cancelling job 4976006 - samtools_subsample_mt
Cancelling job 4976007 - tiddit_coverage
Cancelling job 4976008 - tiddit_coverage
Cancelling job 4976009 - tiddit_coverage
Cancelling job 4976010 - chromograph_cov
Cancelling job 4976011 - chromograph_cov
Cancelling job 4976012 - chromograph_cov
Cancelling job 4976013 - mitodel
Cancelling job 4976014 - mitodel
Cancelling job 4976015 - mitodel
Cancelling job 4976016 - sambamba_depth
Cancelling job 4976017 - sambamba_depth
Cancelling job 4976018 - sambamba_depth
Cancelling job 4976019 - smncopynumbercaller
Cancelling job 4976020 - picardtools_collectmultiplemetrics
Cancelling job 4976021 - picardtools_collectmultiplemetrics
Cancelling job 4976022 - picardtools_collectmultiplemetrics
Cancelling job 4976023 - picardtools_collecthsmetrics
Cancelling job 4976024 - picardtools_collecthsmetrics
Cancelling job 4976025 - picardtools_collecthsmetrics
Cancelling job 4976029 - expansionhunter
Cancelling job 4976031 - tiddit
Cancelling job 4976032 - sv_combinevariantcallsets
Cancelling job 4976033 - sv_annotate
Cancelling job 4976037 - sv_reformat
Cancelling job 4976038 - vcf2cytosure_ar
Cancelling job 4976039 - me_merge_bam
Cancelling job 4976040 - me_merge_bam
Cancelling job 4976041 - me_merge_bam
Cancelling job 4976042 - retroseq
Cancelling job 4976043 - retroseq
Cancelling job 4976044 - retroseq
Cancelling job 4976045 - me_merge_vcfs
Cancelling job 4976046 - me_annotate
Cancelling job 4976047 - me_varianteffectpredictor
Cancelling job 4976048 - me_filter
Cancelling job 4976053 - gatk_combinevariantcallsets
Cancelling job 4976054 - peddy_ar
Cancelling job 4976055 - plink
Cancelling job 4976056 - gatk_variantevalall
Cancelling job 4976057 - gatk_variantevalall
Cancelling job 4976058 - gatk_variantevalall
Cancelling job 4976059 - gatk_collectreadcounts
Cancelling job 4976060 - gatk_collectreadcounts
Cancelling job 4976061 - gatk_collectreadcounts
Cancelling job 4976062 - gatk_denoisereadcounts
Cancelling job 4976063 - gatk_denoisereadcounts
Cancelling job 4976064 - gatk_denoisereadcounts
Cancelling job 4976065 - gens_generatedata
Cancelling job 4976066 - gens_generatedata
Cancelling job 4976067 - gens_generatedata
Cancelling job 4976072 - rhocall_viz
Cancelling job 4976073 - rhocall_viz
Cancelling job 4976074 - rhocall_viz
Cancelling job 4976075 - chromograph_rhoviz
Cancelling job 4976076 - chromograph_rhoviz
Cancelling job 4976077 - chromograph_rhoviz
Cancelling job 4976078 - upd_ar
Cancelling job 4976079 - chromograph_upd
Cancelling job 4976080 - mt_annotation
Cancelling job 4976086 - endvariantannotationblock
Cancelling job 4976087 - gatk_variantevalexome
Cancelling job 4976088 - gatk_variantevalexome
Cancelling job 4976089 - gatk_variantevalexome
Cancelling job 4976090 - qccollect_ar
Cancelling job 4976091 - version_collect_ar
Cancelling job 4976092 - multiqc_ar
Cancelling job 4976093 - analysisrunstatus
Cancelling job 4976094 - sacct
Cancelling job 4975994 - bwa_mem2
Cancelling job 4975993 - bwa_mem2
Cancelling job 4975992 - bwa_mem2
Case justhusky - Analysis 271984: all ongoing jobs cancelled successfully!
Status in SLURM: justhusky - 271984
          id            step      status time_limit  time_elapsed              started
0    4975995  samtools_merge   CANCELLED   10:00:00             0  2023-07-30T15:57:07
1    4975996  samtools_merge   CANCELLED   10:00:00             0  2023-07-30T15:57:07
2    4975997  samtools_merge   CANCELLED   10:00:00             0  2023-07-30T15:57:07
3    4975998  markduplicates   CANCELLED   20:00:00             0  2023-07-30T15:57:07
4    4975999  markduplicates   CANCELLED   20:00:00             0  2023-07-30T15:57:07
..       ...             ...         ...        ...           ...                  ...
101  4975990       fastqc_ar   COMPLETED   10:00:00            42  2023-07-30T14:35:59
102  4975991       fastqc_ar   COMPLETED   10:00:00            42  2023-07-30T14:35:59
103  4975994        bwa_mem2  COMPLETING   10:00:00            73  2023-07-30T14:43:42
104  4975993        bwa_mem2  COMPLETING   10:00:00            80  2023-07-30T14:36:55
105  4975992        bwa_mem2  COMPLETING   10:00:00            80  2023-07-30T14:36:54

[106 rows x 6 columns]
Updated status justhusky - 271984: canceled
```

**Comment by henrikstranneheim** [comment]
### Make analysis failed via GUI ✅ 
Screenshot 2023-07-30 at 16 24 30

### Display info on an analysis ✅ 
Screenshot 2023-07-30 at 16 36 21

**Comment by henrikstranneheim** [comment]
Deployed to prod

**Pull request merged**